src/core/inductive.ml

```diff
@@ -0,0 +1,257 @@
+(** Generating of induction principles. *)
+
+open Timed
+open Pos
+open Console
+open Terms
+open Print
+open Syntax
+
+(** Logging function for generating of inductive principle. *)
+let log_ind = new_logger 'g' "indu" "generating induction principle"
+let log_ind = log_ind.logger
+
+(** Builtin configuration for induction. *)
+type config =
+  { symb_Prop : sym (** : TYPE. Type of propositions. *)
+  ; symb_prf  : sym (** : Prop → TYPE.
+                        Interpretation of propositions as types. *) }
+
+(** [get_config ss pos] build the configuration using [ss]. *)
+let get_config : Sig_state.t -> Pos.popt -> config = fun ss pos ->
+  let builtin = Builtin.get ss pos in
+  { symb_Prop = builtin "Prop"
+  ; symb_prf  = builtin "P" }
+
+(** [change_codom_ind_typ pos ind_sym codom] assumes that the type of
+    [ind_sym] is of the form [Πx1:a1,..., Πxn:an, TYPE]. It then generates
+    a [tbox] similar to this type except that [TYPE] is replaced by
+    [codom [x1;...;xn]]. *)
+let change_codom_ind_typ : popt -> sym -> (tbox list -> tbox) -> tbox =
+  fun pos ind_sym codom ->
+  let rec aux : tvar list -> term -> tbox = fun xs a ->
+    match Basics.get_args a with
+    | (Type, _) -> codom (List.rev_map _Vari xs)
+    | (Prod(a,b), _) ->
+        let (x,b) = Basics.unbind_name b in
+        _Prod (lift a) (Bindlib.bind_var x (aux (x::xs) b))
+    | _ -> fatal pos "The type of %a is not supported" pp_symbol ind_sym
+  in aux [] !(ind_sym.sym_type)
+
+(** [principle ss pos ind_sym cons_list] returns an induction principle which
+    is created thanks to the symbol of the inductive type [ind_sym] (and its
+    position [pos]), its constructors [cons_list] and the signature [ss]. *)
+let principle : Sig_state.t -> popt -> sym -> sym list -> term =
+  fun ss pos ind_sym cons_list ->
+
+  (* STEP 0: Define some tools which will be useful *)
+  let c = get_config ss pos in
+  let p = Bindlib.new_var mkfree "p" in
+  let app  : tbox -> tbox list -> tbox = List.fold_left _Appl in
+  let fapp : sym  -> tbox list -> tbox = fun f ts -> app (_Symb f) ts in
+  let prf_of_p : tbox list -> tbox -> tbox = fun ts t ->
+    fapp c.symb_prf [_Appl (app (_Vari p) ts) t]
+  in
+
+  (* STEP 1: Create the type of the property p *)
+  let prop = _Symb c.symb_Prop in
+  let codom ts = _Impl (fapp ind_sym ts) prop in
+  let p_type = change_codom_ind_typ pos ind_sym codom in
+
+  (* STEP 2: Create each clause according to a constructor *)
+  (* [case_of cons_sym] creates a clause according to a constructor
+     [cons_sym]. *)
+  let case_of : sym -> tbox = fun cons_sym ->
+    let rec aux : tbox list -> term -> tbox = fun xs a ->
+      match Basics.get_args a with
+      | (Symb(s), ts) ->
+          if s == ind_sym then
+            prf_of_p (List.map lift ts) (app (_Symb cons_sym) (List.rev xs))
```

Comment:
Please take `aux: tvar list -> ...` and then use here `List.rev_map _Vari xs` as List.rev_map is tail-recursive. See http://caml.inria.fr/pub/docs/manual-ocaml/libref/List.html.

Reply:
We need to factorize and generalize this aux function (this will be useful in rule generation too):
```
(* [fold_cons_typ codom domrec dom ind_sym cons_sym] generates some value iteratively by going through the structure of [sym_cons.sym_type]. On the codomain, the function [codom] is called on the arguments to which [ind_sym] is applied and the variables of the products (in reverse order). In case of a product, the functions [domrec] and [dom] are called depending on whether the domain type is recursive or not. They are applied to the terms to which [ind_sym] is applied, the variable of the product, and the value built from the codomain. *)
let fold_cons_typ :
      (term list -> tvar list -> 'a)
      -> (term list -> tvar -> 'a -> 'a)
      -> (term list -> tvar -> 'a -> 'a)
      -> sym -> sym -> 'a
  = fun codom domrec dom ind_sym cons_sym ->
  let rec aux xs a =
    match Basics.get_args a with
    | (Symb(s), ts) ->
       if s == ind_sym then codom ts xs
       else fatal pos "%a is not a constructor of %a"
              pp_symbol cons_sym pp_symbol ind_sym
    | (Prod(a,b), _) ->
       let (x,b) = Basics.unbind_name b in
       let b = aux (x::xs) b in
       begin
         match Basics.get_args a with
         | (Symb(s), ts) ->
            if s == ind_sym then domrec ts x b else dom ts x b
         | _ -> fatal pos "The type of %a is not supported" pp_symbol cons_sym
       end
    | _ -> fatal pos "The type of %a is not supported" pp_symbol cons_sym
  in aux [] cons_sym.sym_type
```
Then  `case_of` can be defined as follows:
```
let case_of : sym -> tbox = fun cons_sym ->
  let codom ts xs = prf_of_p (List.map lift ts) (app (_Symb cons_sym) (List.rev_map _Vari xs)) in
  let domrec ts x b =
    let ts = List.map lift ts in
    let a = app (_Symb ind_sym) ts in
    let b = _Impl (prf_of_p ts (_Vari x)) b in
    _Prod a (Bindlib.bind_var x b)
  in
  let dom ts x b =
    let ts = List.map lift ts in
    let a = app (_Symb ind_sym) ts in
    _Prod a (Bindlib.bind_var x b)
  in
  fold_cons_typ codom domrec dom ind_sym cons_sym
```

Reply:
I don't understand why you would like to do that but it's done.

Reply:
As already said, this is to factorize the code of gen_rec_typ and gen_rec_rules.

Reply:
I speak about that (there was a delay with your message):
> Please take `aux: tvar list -> ...` and then use here `List.rev_map _Vari xs` as List.rev_map is tail-recursive. See http://caml.inria.fr/pub/docs/manual-ocaml/libref/List.html.

Why is it better to use `tvar` than `tbox`?

Reply:
See fold_cons_typ.

Reply:
Sorry, I don't understand.
The previous code doesn't work, so I just change a little thing and now it works!
But the structure of the gen_rec_rules isn't the exactly same, please see my email.